# Data Cleaning with Pandas

In [1]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it with `brew install wget`, or use `curl <url> -o <filename>`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the file that you want to unzip.

In [3]:
sales_df = pd.read_csv('data/Real Property Sales.zip')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [4]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [6]:
small_sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [7]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.8+ MB


In [4]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [8]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [5]:
bldg_df.head().T  # some ZipCode starts with 0 so it can change the zipcode 

0                           1  \
Major                                   62304                       62304   
Minor                                    9347                        9378   
BldgNbr                                     1                           1   
NbrLivingUnits                              1                           1   
Address              907  SW 104TH ST   98146  10215   11TH AVE SW  98146   
BuildingNumber                          907                         10215   
Fraction                                                                    
DirectionPrefix                            SW                               
StreetName          104TH                       11TH                        
StreetType                             ST                          AVE      
DirectionSuffix                                                        SW   
ZipCode                                 98146                       98146   
Stories                                     1                           1   
BldgGrade                                   7                           7   
BldgGradeVar                                0                           0   
SqFt1stFloor                             1180                         970   
SqFtHalfFloor                               0                           0   
SqFt2ndFloor                                0                           0   
SqFtUpperFloor                              0                           0   
SqFtUnfinFull                               0                           0   
SqFtUnfinHalf                               0                           0   
SqFtTotLiving                            1570                         970   
SqFtTotBasement                           670                           0   
SqFtFinBasement                           390                           0   
FinBasementGrade                            6                           0   
SqFtGarageBasement                          0                           0   
SqFtGarageAttached                          0                         380   
DaylightBasement                                                        N   
SqFtOpenPorch                             150                           0   
SqFtEnclosedPorch                           0                          80   
SqFtDeck                                    0                           0   
HeatSystem                                  5                           5   
HeatSource                                  2                           2   
BrickStone                                  0                           0   
ViewUtilization                                                         N   
Bedrooms                                    3                           3   
BathHalfCount                               0                           0   
Bath3qtrCount                               0                           1   
BathFullCount                               1                           1   
FpSingleStory                               1                           1   
FpMultiStory                                0                           0   
FpFreestanding                              0                           0   
FpAdditional                                0                           0   
YrBuilt                                  1965                        1969   
YrRenovated                                 0                           0   
PcntComplete                                0                           0   
Obsolescence                                0                           0   
PcntNetCondition                            0                           0   
Condition                                   3                           3   
AddnlCost                                   0                           0   

                                            2                           3  \
Major                                  

In [35]:
bldg_df.columns

Index(['Major', 'Minor', 'BldgNbr', 'NbrLivingUnits', 'Address',
       'BuildingNumber', 'Fraction', 'DirectionPrefix', 'StreetName',
       'StreetType', 'DirectionSuffix', 'ZipCode', 'Stories', 'BldgGrade',
       'BldgGradeVar', 'SqFt1stFloor', 'SqFtHalfFloor', 'SqFt2ndFloor',
       'SqFtUpperFloor', 'SqFtUnfinFull', 'SqFtUnfinHalf', 'SqFtTotLiving',
       'SqFtTotBasement', 'SqFtFinBasement', 'FinBasementGrade',
       'SqFtGarageBasement', 'SqFtGarageAttached', 'DaylightBasement',
       'SqFtOpenPorch', 'SqFtEnclosedPorch', 'SqFtDeck', 'HeatSystem',
       'HeatSource', 'BrickStone', 'ViewUtilization', 'Bedrooms',
       'BathHalfCount', 'Bath3qtrCount', 'BathFullCount', 'FpSingleStory',
       'FpMultiStory', 'FpFreestanding', 'FpAdditional', 'YrBuilt',
       'YrRenovated', 'PcntComplete', 'Obsolescence', 'PcntNetCondition',
       'Condition', 'AddnlCost'],
      dtype='object')

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [36]:
small_bldg_df = bldg_df[['Major', 'Minor','SqFtTotLiving','ZipCode']]

In [37]:
small_bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512114 entries, 0 to 512113
Data columns (total 4 columns):
Major            512114 non-null int64
Minor            512114 non-null int64
SqFtTotLiving    512114 non-null int64
ZipCode          468423 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [38]:
small_sales_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [12]:
small_sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 949922

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [19]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [13]:
small_sales_df['Major'] = pd.to_numeric(small_sales_df['Major'], errors='coerce')

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Did it work?

It worked! Let's do the same thing with the `Minor` parcel number.

In [14]:
small_sales_df['Minor'] = pd.to_numeric(small_sales_df['Minor'], errors='coerce')

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [15]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.8+ MB


Now, let's try our join again.

In [57]:
sales_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

In [40]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [18]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1446234 entries, 0 to 1446233
Data columns (total 6 columns):
Major           1446234 non-null float64
Minor           1446234 non-null float64
DocumentDate    1446234 non-null object
SalePrice       1446234 non-null int64
SqFtDeck        1446234 non-null int64
ZipCode         1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 77.2+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [58]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91  334330.0  1343.0   05/30/2006          0           4600     NaN
92  334330.0  1343.0   05/30/2006          0           4600     NaN
93  334330.0  1343.0   11/26/2001          0           4600     NaN
94  334330.0  1343.0   05/30/2006          0           4600     NaN
95  334330.0  1343.0   06/30/2016          0           4600     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [59]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [60]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1329081 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1329081 non-null float64
Minor            1329081 non-null float64
DocumentDate     1329081 non-null object
SalePrice        1329081 non-null int64
SqFtTotLiving    1329081 non-null int64
ZipCode          1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 71.0+ MB


# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [61]:
sales_data = sales_data[~(sales_data['SalePrice'].isna())]

In [62]:
sales_data.isna().sum()

Major            0
Minor            0
DocumentDate     0
SalePrice        0
SqFtTotLiving    0
ZipCode          0
dtype: int64

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [63]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'].head()

14502    98199-3014
14503    98199-3014
14504    98199-3014
23823    98031-3173
24355    98042-3001
Name: ZipCode, dtype: object

In [64]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1329081 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1329081 non-null float64
Minor            1329081 non-null float64
DocumentDate     1329081 non-null object
SalePrice        1329081 non-null int64
SqFtTotLiving    1329081 non-null int64
ZipCode          1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 71.0+ MB


### 3. Add a column for PricePerSqFt



In [65]:
sales_data['PricePerSqFt']=sales_data.SalePrice/sales_data.SqFtTotLiving

In [66]:
sales_data = sales_data[~(sales_data.PricePerSqFt == 0)]  # drop 0 value

In [68]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
6  423943.0   50.0   07/15/1999      96000            960   98092   
7  423943.0   50.0   01/08/2001     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [70]:
sales_data[sales_data.DocumentDate.str.contains('2019')]

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
379      785996.0   190.0   05/16/2019     808000           3000   98072   
557      367890.0    30.0   03/13/2019     923000           2550   98144   
646      214090.0   110.0   03/06/2019     390000           1780   98042   
905      945920.0   125.0   01/29/2019    1750000           1400   98118   
1084     121800.0   175.0   03/26/2019     556500           1460   98166   
1211     145360.0  2285.0   05/02/2019     690000           1910   98125   
1246     350160.0   125.0   02/06/2019     935000           2460   98117   
1345       4000.0   228.0   04/26/2019     369000           1560   98168   
1463     259772.0    70.0   03/22/2019     525000           2570   98058   
1476      85900.0   115.0   03/21/2019     563750           1690   98106   
1489      22603.0  9195.0   05/07/2019     700000           2200   98177   
1721      45200.0  1890.0   04/01/2019    1085000           1650   98107   
1740     186240.0   105.0   03/19/2019     380000            800   98117   
1895     721740.0   680.0   05/24/2019     780800            860   98122   
2144     352106.0  9028.0   03/21/2019     595000           3580   98022   
2319      30200.0   645.0   03/04/2019     380000           1750   98001   
2651     784920.0   780.0   05/15/2019     405000           1410   98065   
2911     152104.0  9136.0   01/24/2019     268000           1551   98001   
2929     668990.0    20.0   04/30/2019     780000           2320   98155   
3238     117500.0   594.0   02/14/2019     785000           1540   98107   
3510      98400.0   450.0   02/20/2019     409950           1850   98058   
3514     413942.0   460.0   05/22/2019    1630000           3480   98006   
3610     797320.0  2320.0   03/27/2019     540000           1240   98146   
3986     302605.0  9246.0   05/07/2019    1010000           2260   98034   
4195      82007.0  9027.0   02/15/2019     895000           3160   98022   
4597     170840.0   880.0   04/12/2019     515000            890   98108   
4878     882790.0   870.0   02/15/2019     890000           2380   98105   
5041     612390.0    50.0   04/03/2019     458000           1210   98148   
5158     808335.0    40.0   01/09/2019     369950           1660   98055   
5534     937630.0   695.0   03/28/2019     550000           1560   98117   
...           ...     ...          ...        ...            ...     ...   
1439309   47500.0   715.0   01/23/2019    1050000           3180   98107   
1439731  287515.0    20.0   03/26/2019     463000           1220   98106   
1439947  404840.0   365.0   03/21/2019     594000           2160   98059   
1439959  239160.0  1605.0   04/15/2019     760000           1190   98116   
1440315  940672.0    70.0   02/07/2019     549995           2700   98038   
1440638   68640.0   940.0   04/30/2019     777000           2400   98008   
1440661  954510.0   100.0   04/30/2019     974000           3070   98059   
1441188  732756.0   190.0   05/17/2019     856980           3270   98045   
1441220  873180.0   530.0   05/13/2019     382000           2000   98023   
1441286  277660.0   116.0   03/18/2019     740000           1510   98117   
1441288  357480.0   970.0   02/22/2019     595000           2300   98034   
1441359  305270.0   704.0   04/03/2019     839950           1510   98117   
1441705  110800.0   740.0   05/07/2019     308000           1400   98118   
1441933  660022.0    40.0   04/16/2019     825000           2230   98074   
1442083  726620.0    77.0   04/01/2019     740000           1530   98115   
1442166  887900.0    39.0   01/14/2019     301000           1270   98032   
1442379   42505.0  9025.0   05/20/2019       5865           1090   98033   
1442879   92304.0  9521.0   01/17/2019     562500           2450   98168   
1443105  138730.0    80.0   01/18/2019     510000           1390   98011   
1443640  713790.0   840.0   05/07/2019     333000           1350   98092   
1443776  142206.0  9083.0   03/25/2019     580000           

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [71]:
seattle = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]

In [73]:
bool_ind = sales_data.ZipCode.apply(lambda x: True if x in seattle else False)
seattle_data = sales_data[bool_ind]

In [74]:
seattle_data.head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
177  510140.0  4256.0   08/01/2003     235000            870   98115   
178  510140.0  4256.0   03/20/1996     135000            870   98115   
180  510140.0  4256.0   07/09/2013     275000            870   98115   
424  761370.0    75.0   07/29/2014     495000           3410   98105   
426  761370.0    75.0   04/22/2016    1345000           3410   98105   

     PricePerSqFt  
177    270.114943  
178    155.172414  
180    316.091954  
424    145.161290  
426    394.428152

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [76]:
seattle_data[seattle_data.DocumentDate.str.contains('2019')].PricePerSqFt.mean()

520.4716526686619